content/blog/2020-12-16-compcert/index.md

```diff
@@ -0,0 +1,249 @@
+
+## CompCert Structure 
+
+There are 25 passes that comprise CompCert. The first 3 convert C to Clight. There are 19 compiler backend passes at present, 8 of which perform optimizations. The last 3 turn CompCert's internal representation into assembly code which is then linked to generate an executable. Here's the system visualization from the original paper: ![Figure 1](system.png)
+### Frontend
+
+CompCert uses a user-configured preprocessor to interpret macros and preprocessor directives, making no guarantees about the relationship between the preprocessed output and the source code. CompCert uses [Menhir][menhir] to create lexing and parsing automata; lexing is unverified, but parsing has been proven to accept the same language as the C grammar Menhir used to generate the lexer and parser. For instance, if the grammar recognizes [ISO-C-99][c99] (henceforth C99), then the parser automaton is a verified C99 parser. After the C program is parsed into a C AST, an unverified pass typechecks the AST and demarcates reserved registers, linker sections, and definitions to be used in a given section, transforming the C AST into a *CompCert C* AST. 
+
+### Backend
+
+The passes I'll cover are all verified and implemented in Coq (we'll understand what it means to verify a pass in just a few sections!). Feel free to skim or skip this if the system diagram is granular enough. 
+
+The __first pass__ translates CompCert C into Clight by desugaring unsupported expressions and variables. 
+
+The __second pass__ rewrites the Clight program into one that ensures function parameters are temporary variables on the stack at the time the function is called. 
+
+The __third pass__ translates Clight into the [C#minor][csminor] intermediate representation, simplifying loop and `switch` statements in addition to resolving address computation and overloaded function applications. 
+
+The __fourth pass__ translates C#minor into Cminor, where the only addressable variable a function takes is the stack. Cminor is the last intermediate language in the compiler that is independent of computer architecture. Work on [separation logic for Cminor][cmsl] to support reasoning about concurrency as well as the development of a [verified static analyzer for C#minor][csmsa] make Cminor an attractive target for other C compilers. 
+
+The __fifth pass__ translates Cminor into CminorSel, which performs instruction selection on sequences of operations that can be done in one instruction (for instance, an add instruction with a constant becomes an immediate-add instruction for 32-bit Intel architecture). Loads and stores are also translated into the target architecture's addressing conventions. The correctness theorem of this pass does not hold if the target language doesn't support required operations.
```

Comment:
Just to confirm: I assume the implication is not incorrectness but incompleteness, right? That is, the compiler is allowed to say "I refuse to compile this program" but not to silently produce incorrect code in that case?

Reply:
You're right, I'll amend this to reflect that.